### ISSUE TITLE
Failures of test_proxy_connect.py in recent releases

### ISSUE THREAD

**Issue by idella** [scrapy/scrapy#1464]
Under python2.7 (only) in versions 1.0.2 & 1.0.3, test file test_proxy_connect.py yields 4 failures persistently, with the stipulated deps including mitmproxy-0.10.1 and netlib-0.10.1. I installed the plugin pytest-twisted specially for this test since it is not in portage.

Whichever way I ran them I get, under py2.7,

``` bash
scrapy/downloadermiddlewares/ajaxcrawl.py .
scrapy/extensions/httpcache.py .
scrapy/http/cookies.py .
scrapy/utils/datatypes.py .
scrapy/utils/misc.py .
scrapy/utils/python.py ..
scrapy/utils/response.py .
scrapy/utils/template.py .
scrapy/utils/url.py .
tests/test_proxy_connect.py FF..FF

=================== FAILURES ===================
____________ ProxyConnectTestCase.test_https_connect_tunnel ____________

result = None, g = <generator object test_https_connect_tunnel at 0x7f37208f1a00>
deferred = <Deferred at 0x7f37208ec5a8 current result: None>

    def _inlineCallbacks(result, g, deferred):
        """
        See L{inlineCallbacks}.
        """
        # This function is complicated by the need to prevent unbounded recursion
        # arising from repeatedly yielding immediately ready deferreds.  This while
        # loop and the waiting variable solve that by manually unfolding the
        # recursion.

        waiting = [True, # waiting for result?
                   None] # result

        while 1:
            try:
                # Send the last result back as the result of the yield expression.
                isFailure = isinstance(result, failure.Failure)
                if isFailure:
                    result = result.throwExceptionIntoGenerator(g)
                else:
>                   result = g.send(result)

/usr/lib64/python2.7/site-packages/twisted/internet/defer.py:1128: 
 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../work/scrapy-1.0.3/tests/test_proxy_connect.py:56: in test_https_connect_tunnel
    self._assert_got_response_code(200, l)
../../../work/scrapy-1.0.3/tests/test_proxy_connect.py:104: in _assert_got_response_code
    self.assertEqual(str(log).count('Crawled (%d)' % code), 1)
/usr/lib64/python2.7/site-packages/twisted/trial/_synctest.py:437: in assertEqual
    super(_Assertions, self).assertEqual(first, second, msg)
E   FailTest: 0 != 1
------------------- Captured stderr call ---------------
Traceback (most recent call last):
  File "/usr/lib64/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib64/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/mnt/gen2/TmpDir/portage/dev-python/scrapy-1.0.3/work/scrapy-1.0.3/tests/mockserver.py", line 205, in <module>
    httpPort = reactor.listenTCP(8998, factory)
  File "/usr/lib64/python2.7/site-packages/twisted/internet/posixbase.py", line 492, in listenTCP
    p.startListening()
  File "/usr/lib64/python2.7/site-packages/twisted/internet/tcp.py", line 984, in startListening
    raise CannotListenError(self.interface, self.port, le)
twisted.internet.error.CannotListenError: Couldn't listen on any:8998: [Errno 98] Address already in use.
__________ ProxyConnectTestCase.test_https_connect_tunnel_error __________
result = None, g = <generator object test_https_connect_tunnel_error at 0x7f371ef53f00>
deferred = <Deferred at 0x7f371ef825a8 current result: None>

    def _inlineCallbacks(result, g, deferred):
        """
        See L{inlineCallbacks}.
        """
        # This function is complicated by the need to prevent unbounded recursion
        # arising from repeatedly yielding immediately ready deferreds.  This while
        # loop and the waiting variable solve that by manually unfolding the
        # recursion.

        waiting = [True, # waiting for result?
                   None] # result

        while 1:
            try:
                # Send the last result back as the result of the yield expression.
                isFailure = isinstance(result, failure.Failure)
                if isFailure:
                    result = result.throwExceptionIntoGenerator(g)
                else:
>                   result = g.send(result)

/usr/lib64/python2.7/site-packages/twisted/internet/defer.py:1128: 
 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../work/scrapy-1.0.3/tests/test_proxy_connect.py:72: in test_https_connect_tunnel_error
    self._assert_got_tunnel_error(l)
../../../work/scrapy-1.0.3/tests/test_proxy_connect.py:107: in _assert_got_tunnel_error
    self.assertEqual(str(log).count('TunnelError'), 1)
/usr/lib64/python2.7/site-packages/twisted/trial/_synctest.py:437: in assertEqual
    super(_Assertions, self).assertEqual(first, second, msg)
E   FailTest: 0 != 1
----------------------------------------------------- Captured stderr call -----------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib64/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib64/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/mnt/gen2/TmpDir/portage/dev-python/scrapy-1.0.3/work/scrapy-1.0.3/tests/mockserver.py", line 205, in <module>
    httpPort = reactor.listenTCP(8998, factory)
  File "/usr/lib64/python2.7/site-packages/twisted/internet/posixbase.py", line 492, in listenTCP
    p.startListening()
  File "/usr/lib64/python2.7/site-packages/twisted/internet/tcp.py", line 984, in startListening
    raise CannotListenError(self.interface, self.port, le)
twisted.internet.error.CannotListenError: Couldn't listen on any:8998: [Errno 98] Address already in use.
_____ ProxyConnectTestCase.test_https_tunnel_auth_error _____
result = None, g = <generator object test_https_tunnel_auth_error at 0x7f3720c32280>
deferred = <Deferred at 0x7f37259842d8 current result: None>

    def _inlineCallbacks(result, g, deferred):
        """
        See L{inlineCallbacks}.
        """
        # This function is complicated by the need to prevent unbounded recursion
        # arising from repeatedly yielding immediately ready deferreds.  This while
        # loop and the waiting variable solve that by manually unfolding the
        # recursion.

        waiting = [True, # waiting for result?
                   None] # result

        while 1:
            try:
                # Send the last result back as the result of the yield expression.
                isFailure = isinstance(result, failure.Failure)
                if isFailure:
                    result = result.throwExceptionIntoGenerator(g)
                else:
>                   result = g.send(result)

/usr/lib64/python2.7/site-packages/twisted/internet/defer.py:1128: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../work/scrapy-1.0.3/tests/test_proxy_connect.py:82: in test_https_tunnel_auth_error
    self._assert_got_tunnel_error(l)
../../../work/scrapy-1.0.3/tests/test_proxy_connect.py:107: in _assert_got_tunnel_error
    self.assertEqual(str(log).count('TunnelError'), 1)
/usr/lib64/python2.7/site-packages/twisted/trial/_synctest.py:437: in assertEqual
    super(_Assertions, self).assertEqual(first, second, msg)
E   FailTest: 0 != 1
--------- Captured stderr call ------------
Traceback (most recent call last):
  File "/usr/lib64/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib64/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/mnt/gen2/TmpDir/portage/dev-python/scrapy-1.0.3/work/scrapy-1.0.3/tests/mockserver.py", line 205, in <module>
    httpPort = reactor.listenTCP(8998, factory)
  File "/usr/lib64/python2.7/site-packages/twisted/internet/posixbase.py", line 492, in listenTCP
    p.startListening()
  File "/usr/lib64/python2.7/site-packages/twisted/internet/tcp.py", line 984, in startListening
    raise CannotListenError(self.interface, self.port, le)
twisted.internet.error.CannotListenError: Couldn't listen on any:8998: [Errno 98] Address already in use.
________________________ ProxyConnectTestCase.test_https_tunnel_without_leak_proxy_authorization_header ________________________

result = None, g = <generator object test_https_tunnel_without_leak_proxy_authorization_header at 0x7f371eea9410>
deferred = <Deferred at 0x7f372570b758 current result: None>

    def _inlineCallbacks(result, g, deferred):
        """
        See L{inlineCallbacks}.
        """
        # This function is complicated by the need to prevent unbounded recursion
        # arising from repeatedly yielding immediately ready deferreds.  This while
        # loop and the waiting variable solve that by manually unfolding the
        # recursion.

        waiting = [True, # waiting for result?
                   None] # result

        while 1:
            try:
                # Send the last result back as the result of the yield expression.
                isFailure = isinstance(result, failure.Failure)
                if isFailure:
                    result = result.throwExceptionIntoGenerator(g)
                else:
>                   result = g.send(result)

/usr/lib64/python2.7/site-packages/twisted/internet/defer.py:1128: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../work/scrapy-1.0.3/tests/test_proxy_connect.py:91: in test_https_tunnel_without_leak_proxy_authorization_header
    self._assert_got_response_code(200, l)
../../../work/scrapy-1.0.3/tests/test_proxy_connect.py:104: in _assert_got_response_code
    self.assertEqual(str(log).count('Crawled (%d)' % code), 1)
/usr/lib64/python2.7/site-packages/twisted/trial/_synctest.py:437: in assertEqual
    super(_Assertions, self).assertEqual(first, second, msg)
E   FailTest: 0 != 1
----------------------------------------------------- Captured stderr call -----------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib64/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib64/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/mnt/gen2/TmpDir/portage/dev-python/scrapy-1.0.3/work/scrapy-1.0.3/tests/mockserver.py", line 205, in <module>
    httpPort = reactor.listenTCP(8998, factory)
  File "/usr/lib64/python2.7/site-packages/twisted/internet/posixbase.py", line 492, in listenTCP
    p.startListening()
  File "/usr/lib64/python2.7/site-packages/twisted/internet/tcp.py", line 984, in startListening
    raise CannotListenError(self.interface, self.port, le)
twisted.internet.error.CannotListenError: Couldn't listen on any:8998: [Errno 98] Address already in use.
====== 4 failed, most passed in 21.65 seconds ======
```

I was expecting the installing of the plugin of pytest-twisted to make it pass but no impact.

It appears that once the first fails it cascades onto the tests that follow making them fail. Note the 
[Errno 98] Address already. I suspect this is still occupied by a prior test and need be liberated for the following test(s).  Or alternatively are they expected failures that are registered as failures and cause the suite to exit with ====== 4 failed, ???

Do you require any further information?

**Comment by kmike** [comment]
In test_proxy Scrapy starts and stops mockserver before/after each test. So if Scrapy tests are executed sequentially you shouldn't get `Couldn't listen on any:8998: [Errno 98] Address already in use.` errors. Are you by chance running multiple Scrapy test suites in parallel? Or maybe there is another software which uses 8998 port? 

For example, sometimes when you kill test running process manually mockserver is not killed.

We should either randomize host or use empheral ports to prevent such errors (see https://github.com/scrapy/scrapy/pull/1117).

**Comment by idella** [comment]
@kmike ok;
 this testsuite is not being run with  multiple Scrapy test suites in parallel. Definitely one at a time.  No other packages are being run, Just one thing at a time.  What makes me question this is that we have;
tests/test_proxy_connect.py FF..FF
so unless I am missing something the first test run is failing and is therefore not competing with any other test at least from test_proxy_connect.py unless it is tied up from a previous run of the suite. To me they should be cleared on completion of a run whether clean or unclean.  The first test failing then blockades those that follow except that the middle two actually pass. 

It appears the tests are not getting 'clear wind' and so are not actually being run to test what it is they were to test.  I'd say I can exclude these tests for now but they are currently broken or unreliable ( not robust) in the running of the whole suite.

Thx for reply

**Comment by kmike** [comment]
@idella a help with debugging is appreciated because these tests are passing on my local machine (OS X) and in Travis (Ubuntu). 

From error messages I got an idea tests are failing because there is some process which also binds 8998 port on your machine. Scrapy testing suite itself should start and stop mockserver cleanly in case of test failures. So 8998 port should be used either by mockserver not started by current test running process, or by some other software.

Are you sure there are no processes which bind 8998 port on your machine? There is also a possiblity that mock server shutdown code doesn't wait long enough for the port to become free (or that is something os-specific), but let's rule out other reasons first.

>  I'd say I can exclude these tests for now but they are currently broken or unreliable ( not robust) in the running of the whole suite.

You're right that they are unreliable in sense of not being able to run if 8998 port is used. A solution to this is proposed in #1117. Currently to run Scrapy tests one must ensure that 8998 port is free.

**Comment by idella** [comment]
I am not familiar in depth with netstat but

 $ netstat -a | grep 8998
tcp        0      0 0.0.0.0:8998            0.0.0.0:*               LISTEN

$ netstat --numeric-ports | grep 8998 returns empty

So I believe the answer is no 8998 port is not in use

**Comment by cathalgarvey** [comment]
Hi folks, this is marked as not reproducible, and it concerns our tests but doesn't seem to affect the CI. And, it doesn't look like it will give _misleading_ trial results if/when it does occur.

Because of this, and because it's now quite stale, I'm closing it. But I'm open to new contribution here if it still affects anyone on the most recent versions or revisions of Scrapy.
